### PR TITLE
[feat] Add multi label support

### LINKED ISSUE #208: multi-label inconsistencies in the API
## 🐛 Bug

When I try to use the `ImageClassifationData` to solve a multi-label problem there's no clear way how to pass the labels data.

Following the multi-class example in the documentation, I expect the mult-label to work as follows:

```python
train_images: List[str] = ['img1.jpg', 'img2.jpg', ...]
train_labels: List[List[int]] = [[0, 0, 1], [1, 1, 0], ...]

num_classes: int = len(train_labels[0])

datamodule = ImageClassificationData.from_filepaths(
    train_filepaths=train_images,
    train_labels=train_labels,
    batch_size=32,
    num_workers=mproc.cpu_count(),
    train_transform=TRAIN_TRANSFORM,
    val_transform=VALID_TRANSFORM,
    val_split=0.3
)

model = ImageClassifier(
    num_classes=num_classes,
    backbone="resnet50",
    pretrained=True,
    optimizer = torch.optim.Adam,
    loss_fn=torch.nn.functional.binary_cross_entropy_with_logits,
)
```
It seems that the `DataModule` don't cast to torch tensors and don't prepare the batch.

As a workaround, and NOT specified in the docs - you can pass the labels as `List[torch.Tensor]`.
However, you have create and tweak a custom loss function as follows:

```python
def my_loss(x, y):
    # To make this work you have to pass the labels as List[torch.Tensor] from `from_filepaths`.
    # This not specified anywhere, since the examples show that you must pass List[int].
    assert isinstance(x, torch.Tensor), type(x)
    assert isinstance(y, torch.Tensor), type(y)
    assert x.shape == y.shape, (x.shape, y.shape)
    # NOTE: even if you pass the tensor as torch.float it's automatically casted to torch.int64
    y = y.to(x)  # labels have to be in floating point precision to compute BCE
    return torch.nn.functional.binary_cross_entropy_with_logits(x, y)
```

IMO, there should a unique way to pass labels either `List[List[int]]` or `List[torch.Tensor]`.

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR adds support for multi-label

Fixes #208

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [x] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/core/model.py [hunk 1]
```diff
@@ -60,8 +60,8 @@ class Task(LightningModule):
         optimizer: Optimizer to use for training, defaults to `torch.optim.Adam`.
         metrics: Metrics to compute for training and evaluation.
         learning_rate: Learning rate to use for training, defaults to `5e-5`.
-        default_preprocess: :class:`.Preprocess` to use as the default for this task.
-        default_postprocess: :class:`.Postprocess` to use as the default for this task.
+        preprocess: :class:`.Preprocess` to use as the default for this task.
```
Comment (edgarriba) [inline flash/core/model.py:63]:
```suggestion
        preprocess: :class:`Preprocess` to use as the default for this task.
```

File: flash/core/model.py [hunk 2]
```diff
@@ -60,8 +60,8 @@ class Task(LightningModule):
...
+        postprocess: :class:`.Postprocess` to use as the default for this task.
```
Comment (edgarriba) [inline flash/core/model.py:64]:
```suggestion
        postprocess: :class:`Postprocess` to use as the default for this task.
```

File: flash/vision/classification/model.py [hunk 3]
```diff
@@ -24,6 +24,10 @@
 from flash.vision.backbones import IMAGE_CLASSIFIER_BACKBONES
 
 
+def binary_cross_entropy_with_logits(x, y):
```
Comment (edgarriba) [inline flash/vision/classification/model.py:27]:
```suggestion
def binary_cross_entropy_with_logits(x, y):
    """Calls BCE with logits and cast the target one_hot (y) encoding to floating point precision."""
```

File: flash/core/model.py [hunk 4]
```diff
@@ -99,8 +99,7 @@ def step(self, batch: Any, batch_idx: int) -> Any:
         y_hat = self.to_metrics_format(y_hat)
         for name, metric in self.metrics.items():
             if isinstance(metric, torchmetrics.metric.Metric):
-                metric(y_hat, y)
-                logs[name] = metric  # log the metric itself if it is of type Metric
+                logs[name] = metric(y_hat, y)  # log the metric itself if it is of type Metric
```
Comment (ethanwharris) [inline flash/core/model.py:102]: Not logging the metric any more so the comment should change. Is this to do with https://github.com/PyTorchLightning/pytorch-lightning/issues/7052 ? If so, maybe add a todo to change this back once it works again?

File: flash/core/classification.py [hunk 5]
```diff
@@ -22,14 +22,22 @@
...
 
 class ClassificationTask(Task):
 
-    def __init__(self, *args, **kwargs) -> None:
-        super().__init__(*args, default_postprocess=ClassificationPostprocess(), **kwargs)
+    postprocess_cls = ClassificationPostprocess
 
     def to_metrics_format(self, x: torch.Tensor) -> torch.Tensor:
+        if getattr(self.hparams, "multi_label", False):
+            return F.sigmoid(x).int()
```
Comment (ethanwharris) [inline flash/core/classification.py:42]: Some metrics (e.g. AUROC) require just probabilities, maybe casting to int is a bad idea?

**Review by ethanwharris — CHANGES_REQUESTED** [review]

File: flash/vision/classification/model.py [hunk 6]
```diff
@@ -100,4 +114,7 @@ def __init__(
 
     def forward(self, x) -> Any:
```
Comment (edgarriba) [inline flash/vision/classification/model.py:115]:
```suggestion
    def forward(self, x) -> torch.Tensor:
```

File: flash/core/classification.py [hunk 7]
```diff
@@ -11,25 +11,36 @@
...
+from flash.data.process import Postprocess, Preprocess
 
 
 class ClassificationPostprocess(Postprocess):
 
+    def __init__(self, multi_label: bool = False, save_path: Optional[str] = None):
+        super().__init__(save_path=save_path)
+        self.multi_label = multi_label
+
     def per_sample_transform(self, samples: Any) -> Any:
```
Comment (edgarriba) [inline flash/core/classification.py:29]:
```suggestion
    def per_sample_transform(self, samples: Any) -> List[Any]:
```

File: tests/vision/classification/test_model.py [hunk 8]
```diff
@@ -67,3 +79,16 @@ def test_unfreeze():
...
+
+def test_multilabel(tmpdir):
+
+    num_classes = 4
+    ds = DummyMultiLabelDataset(num_classes)
+    model = ImageClassifier(num_classes, multi_label=True)
+    train_dl = torch.utils.data.DataLoader(ds, batch_size=2)
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True)
+    trainer.finetune(model, train_dl, strategy="freeze_unfreeze")
+    image, _ = ds[0]
```
Comment (edgarriba) [inline tests/vision/classification/test_model.py:92]: I'd add test verifying that labels get out as expected. Found issues with that.

File: flash/core/model.py (hunk 4 above)
Comment (tchaton) [reply flash/core/model.py:102]: Changed to what it was before.

**Review by ethanwharris — COMMENTED** [review]
Looks good, see above small comments, meant to click comment rather than request changes haha

File: flash/core/model.py (hunk 1 above)
Comment (ethanwharris) [reply flash/core/model.py:63]: I think this would stop sphinx from generating the links properly?

File: flash/core/classification.py (hunk 5 above)
Comment (tchaton) [reply flash/core/classification.py:45]: Yes, removed !

[the remark this answers is not in the document]
Comment (tchaton) [reply flash/core/classification.py:45]: My bad, it was a bug.

File: flash/core/classification.py [hunk 9]
```diff
@@ -22,14 +22,22 @@
...
-        return torch.argmax(samples, -1).tolist()
+        if self.multi_label:
+            return F.sigmoid(samples).tolist()
```
Comment (aribornstein) [inline flash/core/classification.py:31]: Should we be hardcoding using sigmoid?

Comment (tchaton) [reply flash/core/classification.py:31]: Yes, we need to write it somewhere in the pipeline.

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

**Comment by aribornstein** [comment]
LGTM

**Comment by ThakurRajAnand** [comment]
@tchaton Do we have a multi-label example of doing fine-tuning for merged features?
